### PR TITLE
Gridspec add tight_rect 

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

This adds `set_tight_rect` and `get_tight_rect` to `gridspec.GridspecBase` , and checks the value of this when `gridspec.tight_layout()` is called.  If the `rect` keyword of `tight_layout` is `None`, then `tight_rect` will be used instead.  

### Example

```python
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec

gs = gridspec.GridSpec(3,3)
fig = plt.figure()
for g in gs:
     fig.add_subplot(g)
# we would usually do:
# gs.tight_layout(fig, rect=[0, 0, 0.7, 1.])
gs.set_tight_rect([0, 0, 0.7, 1.])
gs.tight_layout(fig)
# if instead we called:
# gs.tight_layout(fig,rect = [0.2, 0.14, 0.5, 0.5]) 
# this wouldl override the value of tight_rect.
```

### Rationale

The user wouldn't necessarily want to use this function (though they could).  Its for other functions to offer rectangle suggestions to `gridspec.tight_layout()` without the user having to do it manually.  i.e if we create a colorbar that is for more than one axes at `x = 0.85`, then `fig.colorbar` may want to call `gs.set_tight_rect([0, 0, 0.8, 1.])` to constrain `tight_layout` from drawing over the colorbar.  

Note the user can still override the suggestion.  

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/whats_new.rst if major new feature
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Don't understand the codevec/patch failure.

File: lib/matplotlib/gridspec.py [hunk 1]
```diff
@@ -133,9 +136,19 @@ def get_grid_positions(self, fig):
...
+    def get_tight_rect(self):
+        'get the rectangle for use with gridspec.tight_layout'
+        return self._tight_rect
+
+    def set_tight_rect(self, tight_rect):
+        'set a rectangle for use with gridspec.tight_layout'
+        if tight_rect is not None and len(tight_rect) != 4:
+            raise ValueError('Expected keyword rect to be a list or tuple'
+                'with four elements')
+        self._tight_rect =  tight_rect
```
Comment (tacaswell) [inline lib/matplotlib/gridspec.py]: too many spaces after `=`.  This must be one of the files that pep8 does not get run against.

**Comment by tacaswell** [comment]
Can you add a test for this?  It looks like some of it is covered in the test suite, but some of the new lines are not.

**Comment by jklymak** [comment]
@tacaswell  Maybe I was too granular in my pull-request.  This is not supposed to be called by the user, but rather by `colorbar` when it steals space from the gridspec. So I didn't think a test was necessary.  

I think to make things easier, I'll close this PR, and re-open when I have a full working solution put together, including a proper test.  When I did this, I ended up somewhat unsatisfied w/ the results, so I need to think about it all again.  Thanks...

**Comment by tacaswell** [comment]
Even if it is not directly called by users, we are trying to have test coverage of the entire code base.  This is a reasonable discussion to be had about more unit-y test vs integration-y tests, but either way the code should be covered.

If nothing else it means someone else will not accidentally break the intended functionality while trying to do something else (like happens to the unit handling code all the time)!

As always thank you for your work (even if it is not merged yet!).

**Pull request closed without merging**